src/Core/src/Platform/iOS/KeyboardAutoManagerScroll.cs

```diff
@@ -0,0 +1,614 @@
+			nfloat innerScrollValue = 0;
+
+			while (superScrollView is not null)
+			{
+				var shouldContinue = false;
+
+				if (move > 0)
+					shouldContinue = move > -superScrollView.ContentOffset.Y - superScrollView.ContentInset.Top;
+
+				else if (superScrollView.FindResponder<UITableView>() is UITableView tableView)
+				{
+					shouldContinue = superScrollView.ContentOffset.Y > 0;
+
+					if (shouldContinue && View.FindResponder<UITableViewCell>() is UITableViewCell tableCell
+						&& tableView.IndexPathForCell(tableCell) is NSIndexPath indexPath
+						&& tableView.GetPreviousIndexPath(indexPath) is NSIndexPath previousIndexPath)
+					{
+						var previousCellRect = tableView.RectForRowAtIndexPath(previousIndexPath);
+						if (!previousCellRect.IsEmpty)
+						{
+							var previousCellRectInRootSuperview = tableView.ConvertRectToView(previousCellRect, RootController.Superview);
+							move = (nfloat)Math.Min(0, previousCellRectInRootSuperview.GetMaxY() - topLayoutGuide);
+						}
+					}
+				}
+
+				else if (superScrollView.FindResponder<UICollectionView>() is UICollectionView collectionView)
+				{
+					shouldContinue = superScrollView.ContentOffset.Y > 0;
+
+					if (shouldContinue && View.FindResponder<UICollectionViewCell>() is UICollectionViewCell collectionCell
+						&& collectionView.IndexPathForCell(collectionCell) is NSIndexPath indexPath
+						&& collectionView.GetPreviousIndexPath(indexPath) is NSIndexPath previousIndexPath
+						&& collectionView.GetLayoutAttributesForItem(previousIndexPath) is UICollectionViewLayoutAttributes attributes)
+					{
+						var previousCellRect = attributes.Frame;
+
+						if (!previousCellRect.IsEmpty)
+						{
+							var previousCellRectInRootSuperview = collectionView.ConvertRectToView(previousCellRect, RootController.Superview);
+							move = (nfloat)Math.Min(0, previousCellRectInRootSuperview.GetMaxY() - topLayoutGuide);
+						}
+					}
+				}
+
+				else
+				{
+					if (cursorRect.Y - innerScrollValue >= topLayoutGuide && cursorRect.Y - innerScrollValue <= keyboardYPosition)
+						shouldContinue = false;
+					else
+						shouldContinue = true;
+
+					if (cursorRect.Y - innerScrollValue < topLayoutGuide)
+						move = cursorRect.Y - innerScrollValue - (nfloat)topLayoutGuide;
+					else if (cursorRect.Y - innerScrollValue > keyboardYPosition)
+						move = cursorRect.Y - innerScrollValue - keyboardYPosition;
+				}
+
+				// Go up the hierarchy and look for other scrollViews until we reach the UIWindow
+				if (shouldContinue)
+				{
+					var tempScrollView = superScrollView.FindResponder<UIScrollView>();
+					UIScrollView? nextScrollView = null;
+
+					// set tempScrollView to next scrollable superview of superScrollView
+					while (tempScrollView is not null)
+					{
+						if (tempScrollView.ScrollEnabled)
+						{
+							nextScrollView = tempScrollView;
+							break;
+						}
+						tempScrollView = tempScrollView.FindResponder<UIScrollView>();
+					}
```

Comment:
This loop happens several times with a diff generic parameter for the FindResponder, write a generic helper method that does the work for you rather that copy pasting the same loop several times.

Reply:
There are two instances that this happens and they are both UIScrollViews so I didn't make them generic but it is cleaner this way! :)